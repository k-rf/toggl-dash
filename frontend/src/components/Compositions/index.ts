export * from "./Form";
export * from "./RhfTextField";
