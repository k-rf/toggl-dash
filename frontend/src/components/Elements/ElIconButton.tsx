import { IconButton, IconButtonProps } from "@mui/material";

type Props = Pick<
  IconButtonProps,
  "children" | "onClick" | "onMouseDown" | "onMouseUp" | "onMouseLeave" | "disableTouchRipple"
>;

export const ElIconButton = (props: Props) => {
  return <IconButton {...props} />;
};

export type { Props as ElIconButtonProps };
