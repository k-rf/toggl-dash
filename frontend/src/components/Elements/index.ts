export * from "./ElButton";
export * from "./ElPaper";
export * from "./ElTextField";
export * from "./ElAccordion";
