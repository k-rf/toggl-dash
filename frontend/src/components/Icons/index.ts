export * from "./ExpandMoreIcon";
export * from "./PlayArrowIcon";
export * from "./VisibilityIcon";
export * from "./VisibilityOffIcon";
