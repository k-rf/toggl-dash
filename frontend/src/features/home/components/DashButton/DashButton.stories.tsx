import type { Meta, StoryObj } from "@storybook/react";

import { DashButton, DashButtonProps } from "./DashButton";

export default { component: DashButton } as Meta<DashButtonProps>;

type Story = StoryObj<DashButtonProps>;

export const Default: Story = {
  args: {
    summary: "Summary",
    details: "Details",
  },
};

export const Expanded: Story = {
  ...Default,
  args: {
    ...Default.args,
    expanded: true,
  },
};
