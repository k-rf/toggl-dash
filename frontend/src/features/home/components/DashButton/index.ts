export * from "./DashButton";
export * from "./DashButtonDetailListItem";
