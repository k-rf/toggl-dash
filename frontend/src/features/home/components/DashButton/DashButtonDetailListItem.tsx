import { ElListItem } from "~/components/Elements/ElList";
import { ElTypography } from "~/components/Elements/ElTypography";

type Props = {
  label: JSX.Element | string;
  content: JSX.Element | string;
};

export const DashButtonDetailListItem = (props: Props) => {
  return (
    <ElListItem
      sx={{
        p: (theme) => theme.spacing(2, 0, 0),
        flexDirection: "column",
        alignItems: "flex-start",
      }}
    >
      <ElTypography variant="caption">{props.label}</ElTypography>
      {props.content}
    </ElListItem>
  );
};
