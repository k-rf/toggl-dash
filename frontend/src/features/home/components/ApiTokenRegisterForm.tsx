import { InputAdornment, Stack } from "@mui/material";
import { KeyboardEventHandler, MouseEventHandler, useState } from "react";

import { Form, HandleSubmit, Methods } from "~/components/Compositions";
import { ElButton, ElTextField } from "~/components/Elements";
import { ElIconButton } from "~/components/Elements/ElIconButton";
import { ElTypography } from "~/components/Elements/ElTypography";
import { VisibilityIcon, VisibilityOffIcon } from "~/components/Icons";
import { z } from "~/lib/zod";

const schema = z.object({
  apiToken: z.string().min(1),
});

type ApiTokenValues = z.infer<typeof schema>;

type Props = { onSubmit?: (value: ApiTokenValues) => Promise<void> | void };

export const ApiTokenRegisterForm = (props: Props) => {
  const [visible, setVisible] = useState(false);

  const submit = (methods: Methods<ApiTokenValues>) => {
    return methods.handleSubmit(async (value) => {
      methods.reset({ apiToken: "" });

      await props.onSubmit?.(value);
    });
  };

  const handleSubmit: HandleSubmit<ApiTokenValues, MouseEventHandler> = (methods) => {
    return submit(methods);
  };

  const handleKeyDown: HandleSubmit<ApiTokenValues, KeyboardEventHandler> = (methods) => {
    return (e) => {
      if (e.code === "Enter") {
        e.preventDefault();

        if (e.ctrlKey) {
          return submit(methods)(e);
        }
      }
    };
  };

  const handleShow = () => {
    setVisible(true);
  };

  const handleMask = () => {
    setVisible(false);
  };

  return (
    <Form<ApiTokenValues> options={{ mode: "all" }} schema={schema}>
      {(methods) => {
        const { error } = methods.getFieldState("apiToken");
        const {
          errors: { apiToken },
          isValid,
        } = methods.formState;

        return (
          <Stack spacing={1}>
            <ElTypography>Toggl の API トークンを入力してください</ElTypography>
            <ElTextField
              fullWidth
              placeholder="API トークン"
              type={visible ? "text" : "password"}
              error={Boolean(error)}
              onKeyDown={handleKeyDown(methods)}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ElIconButton
                      disableTouchRipple
                      onMouseDown={handleShow}
                      onMouseUp={handleMask}
                      onMouseLeave={handleMask}
                      data-testid="api-token-register-form-visibility-toggle-icon-button"
                    >
                      {visible ? <VisibilityOffIcon /> : <VisibilityIcon />}
                    </ElIconButton>
                  </InputAdornment>
                ),
              }}
              registration={methods.register("apiToken")}
            />
            <ElButton
              disabled={!isValid}
              onClick={handleSubmit(methods)}
              data-testid="api-token-register-form-submit-button"
            >
              保存する
            </ElButton>
            <ElTypography variant="body2" color="error">
              {apiToken?.message}
            </ElTypography>
          </Stack>
        );
      }}
    </Form>
  );
};

export type { Props as ApiTokenRegisterFormProps };
