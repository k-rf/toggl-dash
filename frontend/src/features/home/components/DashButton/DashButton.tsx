import { Box } from "@mui/material";

import {
  ElAccordion,
  ElAccordionDetails,
  ElAccordionSummary,
} from "~/components/Elements/ElAccordion";
import { ElIconButton } from "~/components/Elements/ElIconButton";
import { ExpandMoreIcon, PlayArrowIcon } from "~/components/Icons";

type Props = {
  summary: React.ReactNode;
  details: React.ReactNode;
  expanded?: boolean;
  onChange: () => void;
  onStart: () => void;
};

export const DashButton = (props: Props) => {
  return (
    <ElAccordion
      expanded={props.expanded}
      sx={{ minWidth: 240, maxWidth: 640 }}
      data-testid="dash-button-accordion"
    >
      <ElAccordionSummary>
        {props.summary}
        <Box flexGrow={1} />
        <ElIconButton
          onClick={(e) => {
            e.preventDefault();
            e.stopPropagation();

            props.onStart();
          }}
          color="success"
        >
          <PlayArrowIcon />
        </ElIconButton>
        <ElIconButton
          onClick={(e) => {
            e.preventDefault();
            e.stopPropagation();

            props.onChange();
          }}
          sx={{
            transform: `rotate(${props.expanded ? 180 : 0}deg)`,
            transition: "transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;",
          }}
          data-testid="dash-button-expansion-button"
        >
          <ExpandMoreIcon />
        </ElIconButton>
      </ElAccordionSummary>
      <ElAccordionDetails sx={{ maxHeight: 320, overflow: "auto" }}>
        {props.details}
      </ElAccordionDetails>
    </ElAccordion>
  );
};

export type { Props as DashButtonProps };
