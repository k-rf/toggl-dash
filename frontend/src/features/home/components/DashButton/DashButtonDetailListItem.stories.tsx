import type { Meta, StoryObj } from "@storybook/react";

import { DashButtonDetailListItem } from "./DashButtonDetailListItem";

export default { component: DashButtonDetailListItem } as Meta<typeof DashButtonDetailListItem>;

type Story = StoryObj<typeof DashButtonDetailListItem>;

export const Default: Story = {
  args: {
    label: "Detail Label",
    content: "Detail Content",
  },
};
