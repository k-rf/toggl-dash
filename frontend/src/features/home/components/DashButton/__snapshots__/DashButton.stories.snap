// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/Home/Components/DashButton/DashButton > Default 1`] = `
<body
  style=""
>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded css-shs878-MuiPaper-root-MuiAccordion-root"
      data-testid="dash-button-accordion"
    >
      <div
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root css-10i7duh-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content css-1betqn-MuiAccordionSummary-content"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-16rr9ae-MuiTypography-root"
          >
            Summary
          </p>
          <div
            class="MuiBox-root css-i9gxme"
          />
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeMedium css-1r6h6t4-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PlayArrowIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8 5v14l11-7z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1w456y2-MuiButtonBase-root-MuiIconButton-root"
            data-testid="dash-button-expansion-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-402ff7-MuiAccordionDetails-root"
              >
                <hr
                  class="MuiDivider-root MuiDivider-fullWidth css-1winrj1-MuiDivider-root"
                />
                Details
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Features/Home/Components/DashButton/DashButton > Expanded 1`] = `
<body
  style=""
>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded css-shs878-MuiPaper-root-MuiAccordion-root"
      data-testid="dash-button-accordion"
    >
      <div
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded css-10i7duh-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content Mui-expanded css-1betqn-MuiAccordionSummary-content"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-16rr9ae-MuiTypography-root"
          >
            Summary
          </p>
          <div
            class="MuiBox-root css-i9gxme"
          />
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeMedium css-1r6h6t4-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PlayArrowIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8 5v14l11-7z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-nvinfh-MuiButtonBase-root-MuiIconButton-root"
            data-testid="dash-button-expansion-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-402ff7-MuiAccordionDetails-root"
              >
                <hr
                  class="MuiDivider-root MuiDivider-fullWidth css-1winrj1-MuiDivider-root"
                />
                Details
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
